[1단계 - 로또 구현] 벨로(방예혁) 미션 제출합니다.

YehyeokBang opened the pull request:
안녕하세요 리뷰어님! 7기 백엔드 크루 벨로라고 합니다. 

현실에서 로또를 구매하는 것처럼 캐셔, 로또 발급기 객체들을 활용하여 구현해봤습니다. 잘 부탁드려요! 🙇

### 느낀점

비교적 짧은 시간 안에 페어와 생각을 나누며 코드를 작성하는 것이 꽤나 어려운 것 같아요. 그래도 혼자 구현하는 것보다 많은 생각을 하게 되어서 페어 미션에 매력을 느낀 것 같습니다.

### 궁금한 점

- 출력을 위한 포맷 문자열을 상수로 선언해보려고 했으나, 하나의 메시지 포맷은 보통 하나의 메서드에서만 사용하니까 변경 용이성에 대해서 장점을 느끼지 못했고, 오히려 해당 메서드에서 어떤 형태로 출력이 되는지 판단하기 어려웠던 것 같아요. 리뷰어님의 상수 처리의 간단한 기준이 있으시다면 의견이 궁금합니다!
- 1~6 숫자를 가진 로또를 생성하는 코드를 여러 테스트에 작성하게 되었습니다. 테스트 코드 작성을 할 때 비슷한 객체를 동일하게 생성하는 것을 따로 분리해도 괜찮을까요? (테스트 코드에 익숙치 않아 어렵네요..)
- 함께 작성하는 코드이기 때문에 메서드 분리와 메서드 이름을 잘 짓기 위해 노력했습니다. 기타 개선 사항이나 어색한 부분이 느껴지시면 말씀 부탁드립니다.

시간 내어 리뷰해주셔서 감사합니다. 👍

rladmstn committed:
docs: 구현할 기능 목록 작성

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

rladmstn committed:
feat: 구입 금액 입력 요청 기능 구현

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

rladmstn committed:
feat: 구입금액 음수 검증 기능 구현

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

rladmstn committed:
feat: 로또 발급 기능 구현

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

rladmstn committed:
test: 구입 금액 단위 테스트 작성

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

rladmstn committed:
feat: 구매한 로또 출력 기능 구현

Co-authored-by: YehyeokBang <qkddpgur318@gmail.com>

rladmstn committed:
feat: 당첨 번호 입력 기능 구현

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

rladmstn committed:
feat: 보너스 번호 입력 기능 구현

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

rladmstn committed:
feat: 로또 당첨 통계 출력 기능 구현

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

rladmstn committed:
test: 로또 관련 테스트 작성

Co-authored-by: YehyeokBang qkddpgur318@gmail.com

YehyeokBang committed:
feat: 콘솔 입력 기능 분리

Co-authored-by: rladmstn <oz33177351@gmail.com>

YehyeokBang committed:
refactor: 패키지 및 메서드 분리

Co-authored-by: rladmstn <oz33177351@gmail.com>

rladmstn committed:
refactor: 메서드 분리

Co-authored-by: YehyeokBang <qkddpgur318@gmail.com>

rladmstn committed:
refactor: 들여쓰기 요구사항 반영

Co-authored-by: YehyeokBang <qkddpgur318@gmail.com>

YehyeokBang committed:
refactor: 당첨 순위 관련 리팩토링

- 당첨 확인 로직을 단순화합니다.
- 당첨 확인과 당첨 내역 관리를 분리합니다.

Co-authored-by: rladmstn <oz33177351@gmail.com>

YehyeokBang committed:
test: 테스트 도구 통일

- 혼용해서 사용하던 것을 통일합니다. AssertJ → JUnit

Co-authored-by: rladmstn <oz33177351@gmail.com>

YehyeokBang committed:
refactor: 출력 관련 코드 리팩토링

Co-authored-by: rladmstn <oz33177351@gmail.com>

YehyeokBang committed:
refactor: 패키지 구조 리팩토링

Co-authored-by: rladmstn <oz33177351@gmail.com>

YehyeokBang committed:
refactor: 메서드 분리 및 코드 정렬

Co-authored-by: rladmstn <oz33177351@gmail.com>

rladmstn committed:
refactor: 로또 번호 출력 리팩토링

Co-authored-by: YehyeokBang <qkddpgur318@gmail.com>

rladmstn committed:
test: 로또 발급 테스트 작성

Co-authored-by: YehyeokBang <qkddpgur318@gmail.com>

Hyunta commented on README.md:
```diff
@@ -1,3 +1,29 @@
 # java-lotto
 
 로또 미션 저장소
+
+간단한 로또 발매 프로그램을 구현합니다.
+
+## 기능 목록
+
+- [x] 구입금액을 입력 받는다.
+    - [x] `구입금액을 입력해 주세요.` 문구를 출력한다.
+    - [x] `1,000원` 단위여야 한다.
+    - [X] 음수일 수 없다.
+- [x] 구매한 로또를 출력한다.
+    - [x] `n개를 구매했습니다.` 문구를 출력한다.
+    - [x] 로또 번호는 오름차순으로 정렬한다.
+    - [x] 로또 번호는 대괄호로 감싸고 쉼표로 구분한다.
+- [x] 지난 주 당첨 번호를 입력 받는다.
+    - [x] `지난 주 당첨 번호를 입력해 주세요.` 문구를 출력한다.
+    - [x] 1~45 사이의 수여야 한다.
+    - [x] 6개의 중복되지 않는 수여야 한다.
+    - [x] 음수일 수 없다.
+- [x] 보너스 볼을 입력 받는다.
+    - [x] `보너스 볼을 입력해 주세요.` 문구를 출력한다.
+    - [x] 1~45 사이의 수여야 한다.
+    - [x] 지난 주 당첨 번호와 중복되지 않는 수여야 한다.
+    - [x] 음수일 수 없다.
+- [x] 당첨 통계를 출력한다.
+    - [x] 당첨 결과를 출력한다.
+    - [x] 총 수익률을 출력한다.
```
기능 명세 정리를 잘 해주셨네요 👍🏻

Hyunta commented on src/test/java/lotto/LottoTest.java:
```diff
@@ -0,0 +1,72 @@
+import org.junit.jupiter.params.provider.ValueSource;
+
+class LottoTest {
+
```
커밋 내역으로 보았을 때, 필요한 기능을 전부 개발한 뒤에 한꺼번에 테스트를 작성해주신 것 같네요.
기능을 개발할 때 같이 테스트하지 않고 일괄적으로 만드신 이유가 있을까요?

YehyeokBang replied:
저는 `실제 코드`와 `테스트`가 함께 있어야 기능을 구현했다고 생각했습니다. 그래서 평소에는 구현과 테스트를 함께 커밋하는 방식을 선호했습니다.

하지만 이번 미션에서는 제 졸업식 일정으로 인해 페어와 함께 작업할 시간이 부족했습니다. 😢 
페어 프로그래밍이 아직 익숙하지 않아, 마감 기한을 맞추기 위해 먼저 동작하는 코드를 작성한 후, 시간이 남을 경우 테스트 코드를 추가하는 방식으로 진행하게 되었습니다. 점진적으로 테스트를 추가하며 개선하는 경험을 해보고 싶었지만, 일정상의 이유로 우선 돌아가는 코드를 만드는 것을 선택했습니다.

테스트로 검증하는 것이 아니라, 될 것이라고 믿는 코드를 작성하다 보니 문제 지점을 찾을 때 많은 시간이 걸렸던 것 부분도 아쉬웠습니다. 다음에는 테스트도 함께 작성하는 방법을 시도해 보려고 합니다!

Hyunta replied:
일정이 많이 타이트 하셨군요ㅠㅠ
그쵸 경우에 따라서는 테스트코드를 함께 만들 시간이 모자랄 수 있습니다. 그럼에도 불구하고 테스트코드까지 작성해서 잘 제출해주셨네요 👍🏻

Hyunta commented on src/test/java/lotto/model/LottoTest.java:
```diff
@@ -0,0 +1,72 @@
+
+class LottoTest {
+
+    @DisplayName("중복 없는 6개의 숫자로 로또를 생성할 수 있다.")
+    @Test
+    void ok() {
+        Set<Integer> numbers = Set.of(1, 2, 3, 4, 5, 6);
+
+        assertDoesNotThrow(() -> new Lotto(numbers));
+    }
```
테스트의 이름이 조금 더 구체적이면 어떨까요?

YehyeokBang replied:
다른 예외 발생 테스트 코드처럼 성공(올바르게 수행되는)하는 테스트 코드도 이름을 명확하게 지어주면 좋을 것 같네요! 
감사합니다. 👍

Hyunta commented on src/test/java/lotto/model/LottoMachineTest.java:
```diff
@@ -0,0 +1,25 @@
+    void ok_DistinctNumbers() {
+        Lotto lotto = LottoMachine.issue();
+
+        assertThat(lotto.getNumbers().stream().distinct().count()).isEqualTo(lotto.getNumbers().size());
```
한 줄에 `.` 이 지나치게 많으면 가독성을 해칠 수 있어요.
변수로 추출해내면 어떨까요?

YehyeokBang replied:
아서의 말씀처럼 지금 보니 `무엇이 같은지`가 코드에 명확히 표현되지 않는 것 같네요. 수정해보겠습니다! 👍

Hyunta commented on README.md:
```diff
@@ -1,3 +1,29 @@
 # java-lotto
 
 로또 미션 저장소
+
```
# 질문에 대한 답변

> 출력을 위한 포맷 문자열을 상수로 선언해보려고 했으나, 하나의 메시지 포맷은 보통 하나의 메서드에서만 사용하니까 변경 용이성에 대해서 장점을 느끼지 못했고, 오히려 해당 메서드에서 어떤 형태로 출력이 되는지 판단하기 어려웠던 것 같아요. 리뷰어님의 상수 처리의 간단한 기준이 있으시다면 의견이 궁금합니다!

우선 저 또한 상수로 분리하는 기준은 없습니다.
아주 좋은 경험을 해주신 것 같은데, AB 방법에 대한 고민이 있을 때 간단하게 구현해서 직접 비교해보는 것도 좋은 방법입니다.
상수로 분리할까 말까 고민이 됐을 때 직접 분리를 해보고 오히려 불편해졌다면 합치면 됩니다 간단하죠ㅎㅎ

보통 무언가를 분리를 할 때 말해주신 것처럼 변경에 유연하고, 재사용성을 챙기기 위해서 분리를 하게 됩니다.
하지만 항상 장점만 있지는 않아요 말해주신 것처럼 코드가 흩어지기 때문에 오히려 가독성은 떨어질 수 있습니다.
그렇기 때문에 상황에 맞춰서 적절하게 분리를 해주는 것이 중요한데 이것은 저는 경험을 통해서 채워지는 영역이라고 생각합니다.

> 1~6 숫자를 가진 로또를 생성하는 코드를 여러 테스트에 작성하게 되었습니다. 테스트 코드 작성을 할 때 비슷한 객체를 동일하게 생성하는 것을 따로 분리해도 괜찮을까요? (테스트 코드에 익숙치 않아 어렵네요..)

그쵸 테스트를 만들다보면 누구나 그런 고민에 한번 빠지게 되는 것 같습니다.
스스로 어떻게 할지도 고민해보시고, 남들은 어떻게 할까 궁금하시면 Test Fixture 라는 용어를 찾아보시면 도움이 되실 것 같네요ㅎㅎ

> 함께 작성하는 코드이기 때문에 메서드 분리와 메서드 이름을 잘 짓기 위해 노력했습니다. 기타 개선 사항이나 어색한 부분이 느껴지시면 말씀 부탁드립니다.

페어 프로그래밍을 하다보면 정말 생각지도 못한 부분에서 논의해야할 것들이 생기더라구요.
2단계를 진행하면서 이 부분도 같이 피드백 드려보겠습니다!

YehyeokBang replied:
코드를 작성할 때 정답이 없기 때문에 항상 잘하고 있는가?에 대한 걱정이 드는 것 같아요.

말씀해주신 것처럼 상황에 따라 다르기 때문에 스스로 판단을 했다면 그것을 믿고 작성하고, 여러 사람들과 의견을 공유하며 재정립해나가는 과정이 더욱 중요할 것 같네요.

아서의 답변이 너무 친절하네요! 감사합니다. 👍 👍 👍 
피드백 수정 후에도 잘 부탁드립니다!

Hyunta replied:
우테코 기간동안 본인만의 철학을 세워간다는 생각으로 뭐가 더 좋은 코드일지 고민해보는 습관을 가지면 도움이 되실 것 같습니다ㅎㅎ

Hyunta reviewed (changes requested):
안녕하세요 벨로,
우테코 7기에 합류하게 되신걸 축하드립니다! 🥳
첫 미션 리뷰어로 만나게 된 아서라고 합니다.

온보딩 미션 1단계는 미션 사이클을 경험하시는 것이 주된 목표이기 때문에 코드에 대한 리뷰보다는 전반적인 컨벤션, 커밋 로그, 문서 관리, 테스트와 같은 부분을 중점으로 리뷰를 진행했습니다.
2단계 진행하면서 코드에 대한 부분들도 리뷰를 해드릴테니 너무 걱정하지는 마세요.

프리코스를 이미 한번 경험하셔서인지 커밋 로그도 잘 남겨주셨고, 테스트도 적절하게 진행해주셨습니다.
지금 습관을 잘 유지해주셔서 남은 미션도 잘 진행하시면 많이 배워가실 수 있을 것 같습니다.
언제든지 궁금한 부분이 있으면 DM 주시거나 커멘트로 남겨주세요.

YehyeokBang committed:
refactor: 구매 금액 검증 관련 코드 리팩토링

- 0원인지 확인하는 로직을 추가하여 최소 주문 개수가 1개 이상일 수 있도록 합니다.
- 구매 금액으로 개수를 계산하는 메서드에서 검증 로직 부분을 메서드로 분리합니다.
- 테스트 메서드명을 명확하게 수정합니다.

YehyeokBang committed:
feat: 자주 사용되는 로또 객체 생성을 도와주는 Fixtures 객체 추가

- 1~6 번호를 가진 로또 생성 메서드를 추가합니다.
- 주어진 숫자로 로또를 생성하는 메서드를 추가합니다.
- 테스트 메서드명을 명확히하고 비교하려는 값을 변수로 추출합니다.

YehyeokBang committed:
refactor: 테스트 메서드명 및 표시 이름 수정

YehyeokBang committed:
feat: 순위로 당첨 수를 알 수 있는 메서드 추가

- 기존 getter로 순위 map을 꺼내서 외부에서 키로 찾는 방식이었습니다.
- 키를 전달하여 당첨 수를 반환하는 메서드를 추가합니다.

YehyeokBang committed:
refactor: 테스트 표시명 변경

- Fixture 객체를 사용하도록 변경합니다.

YehyeokBang commented:
```java
assertThat(lotto.getNumbers().stream().distinct().count()).isEqualTo(lotto.getNumbers().size());
-> 
assertThat(countOfIssuedLottoNumbers).isEqualTo(distinctLottoNumberCount);
```

- 위와 같이 `.`이 지나치게 많아 무엇을 비교하는지 파악하기 힘든 코드를 변수로 추출하여 무엇을 비교하는지 명확하게 작성했습니다.
- 테스트 코드를 수정하다가 구입 금액이 0원인 경우에도 프로그램이 정상적으로 작동하는 것을 확인했습니다.
  - 실제 직원에게 0원을 건내며 로또를 사겠다는 일은 발생하지 않으니 예외 처리를 추가해봤습니다.
- 1~6으로 구성된 로또를 생성하거나 직접 숫자를 전달하여 로또를 생성하는 메서드를 가진 Fixture 객체를 추가해봤습니다.  
  - 추가하면서 느낀 장점은 테스트를 위해 어떤 로또 객체가 생성되는지 더 명확하게 표현할 수 있다는 점인 것 같아요. (`1~6로 구성된 로또`, `전달된 숫자들로 만들어진 로또`)
  - 사실 가독성 향상을 목적으로 도입했지만, 큰 변화는 없었던 것 같아요. 로또 객체가 생성하기 간단한 편이기도 하고, 이미 로또라는 객체는 익숙해서 일까요? new로 생성하는 방법일 때도 큰 무리없이 이해할 수 있었던 것 같아요.
  - 그러나 예를 들어, 주문의 객체를 생성할 때 내부적으로 상품, 사용자, 결제 수단 등 여러 객체가 포함되어야 할 때 Fixture 객체를 따로 생성하면 테스트 코드 가독성 향상에 많은 도움이 될 것 같아요.
- 전체적으로 테스트 코드의 메서드명을 명확하게 다듬었습니다!
- 빼먹었던 프롤로그의 [셀프체크 링크](https://prolog.techcourse.co.kr/studylogs/3893)도 추가하겠습니다..!

잘 부탁드립니다! 아서! 👍

Hyunta reviewed (approved):
안녕하세요 벨로,
피드백 반영 잘 해주셨네요!
이번 단계에서는 코드보다는 미션을 진행하는 방식에 대한 피드백을 많이 드렸습니다.
2단계 리팩터링을 진행하시고 다시 또 이야기 나누시죠.
